packages/react/src/components/Dropdown/Dropdown.tsx

```diff
@@ -616,12 +632,42 @@ class Dropdown extends AutoControlledComponent<Extendable<DropdownProps>, Dropdo
   }
 
   private handleStateChange = (changes: StateChangeOptions<ShorthandValue>) => {
+    // console.log(changes)
     if (changes.isOpen !== undefined && changes.isOpen !== this.state.open) {
-      this.trySetStateAndInvokeHandler('onOpenChange', null, { open: changes.isOpen })
+      const newState = { open: changes.isOpen }
+
+      if (changes.isOpen) {
+        if (changes.type === Downshift.stateChangeTypes.keyDownArrowUp) {
+          newState['highlightedIndex'] = _.isNumber(this.state.highlightedIndex)
+            ? this.state.highlightedIndex - 1
+            : this.props.items.length - 1
+        }
+        if (changes.type === Downshift.stateChangeTypes.keyDownArrowDown) {
+          newState['highlightedIndex'] = _.isNumber(this.state.highlightedIndex)
+            ? this.state.highlightedIndex + 1
+            : 0
+        }
+        if (!this.props.search) {
+          this.listRef.current.focus()
+        }
+      } else {
+        if (!this.props.multiple && !this.props.search) {
+          if (changes.selectedItem) {
+            newState['highlightedIndex'] = this.props.items.indexOf(changes.selectedItem)
+          } else if (this.state.value) {
+            newState['highlightedIndex'] = this.props.items.indexOf(this.state.value)
+          }
+        } else {
+          newState['highlightedIndex'] = this.props.resetHighlightedIndex || null
+        }
+      }
```

Comment:
Sorry, but I lost there 😴 If you will try to cover this code with UTs you will not survive 🦇 

1. Can we extract code that calculates only `highlightedIndex` based on changes/props?
2. Can we dispatch DOM effects in `componentDidUpdate()`? I.e. in a single place.
```jsx
componentDidUpdate(prevProps, prevState) {
  if (!prevState.open && this.state.open) {
    if (search) {
      this.listRef.current.focus()
    }
    if (multiple) {
      this.triggerRef.current.focus()
    }
  }
}
```

Reply:
I extracted the code, you can take a look. as for the DOM effects, it's out of scope, but we can create a separate issue. I also want to move the methods around in the Dropdown, since it's kind of chaos there. should be render, then renderWhatevers, then event handlers, then helpers. Or something like that, but grouped better, since it's a 1100 line file.